### ISSUE TITLE
Rocket support?

### ISSUE THREAD

**Issue by numToStr** [async-graphql/async-graphql#222]
Is [rocket](https://github.com/SergioBenitez/Rocket) support planned? As rocket can be built on stable rust.

I really love this lib and using this with rocket will be awesome.

**Comment by sunli829** [comment]
In the plan, it should be added in the next few days.

**Comment by sunli829** [comment]
I tried it. `Rocket` does not currently support stable, so this work needs to be postponed. If you are familiar with `Rocket`, would you like to contribute an `async-graphql-rocket`? 😁

**Comment by numToStr** [comment]
I am just getting started with rust and I was trying out rocket. Maybe I'll just wait for the `Rocket` support. 🤘🏻

**Comment by nicolaiunrein** [comment]
Isn’t Rocket a synchronous framework?

**Comment by numToStr** [comment]
I read that Rocket 0.5 will be async

**Comment by Weasy666** [comment]
I'm currently trying to implement this [here](https://github.com/simplificAR/async-graphql/tree/master/async-graphql-rocket), but...i hit this [problem](https://gist.github.com/Weasy666/2a230114b74aed7a06a7cb90cab6aeb3).
I think the problem could be the combination of generics and the Rockets route macros, but i'm not sure. Need to investigate some more, but wanted to tell that i'm working on this.

**Comment by sunli829** [comment]
> I'm currently trying to implement this [here](https://github.com/simplificAR/async-graphql/tree/master/async-graphql-rocket), but...i hit this [problem](https://gist.github.com/Weasy666/2a230114b74aed7a06a7cb90cab6aeb3).
> I think the problem could be the combination of generics and the Rockets route macros, but i'm not sure. Need to investigate some more, but wanted to tell that i'm working on this.

Is it possible to avoid using generics by defining new types?
```rust
type MySchema = Schema<.....>;
```

**Comment by Weasy666** [comment]
Don't we need generics, for people to be able to provide their own schema?

**Comment by sunli829** [comment]
> Don't we need generics, for people to be able to provide their own schema?

Sorry, I don’t know how to use rocket. 😂
This seems to be a problem with Rocket.

**Comment by Weasy666** [comment]
Ok...it is compiling, see [custom_handler](https://github.com/simplificAR/async-graphql/tree/custom_handler/async-graphql-rocket) branch on my fork.

But...i ran into another problem, while implementing the [Starwars example](https://github.com/simplificAR/examples/tree/master/rocket/starwars). In the [`main.rs`](https://github.com/simplificAR/examples/blob/master/rocket/starwars/src/main.rs) file on line 7 i get the error
```  rust
error[E0277]: the trait bound `starwars::QueryRoot: async_graphql::ObjectType` is not satisfied
 --> rocket\starwars\src\main.rs:7:18
  |
7 |     let schema = Schema::build(QueryRoot, EmptyMutation, EmptySubscription)
  |                  ^^^^^^^^^^^^^ the trait `async_graphql::ObjectType` is not implemented for `starwars::QueryRoot`
  |
  = note: required by `async_graphql::Schema::<Query, Mutation, Subscription>::build`
```

and on line 8 i get the error (but i guess that this one is a consequence of the previous error)
``` rust
error[E0599]: no method named `data` found for struct `async_graphql::SchemaBuilder<starwars::QueryRoot, async_graphql::EmptyMutation, async_graphql::EmptySubscription>` in the current scope
  --> rocket\starwars\src\main.rs:8:10
   |
8  |         .data(StarWars::new())
   |          ^^^^ private field, not a method
   | 
  ::: E:\simplificAR\examples\models\starwars\src\model.rs:88:1
   |
88 | pub struct QueryRoot;
   | --------------------- doesn't satisfy `starwars::QueryRoot: async_graphql::ObjectType`
   |
   = note: the method `data` exists but the following trait bounds were not satisfied:
           `starwars::QueryRoot: async_graphql::ObjectType`

error: aborting due to 2 previous errors
```

I'm new to `async-graphql`, so maybe i'm doing something wrong? 🤔

**Comment by sunli829** [comment]
Because the Async-graphql version is used differently.

https://github.com/simplificAR/examples/blob/22c959f8ee714eb24c9550c6d6b569b2dbc5a2b5/models/starwars/Cargo.toml#L8

**Comment by Weasy666** [comment]
Do you mean that this is a difference between 0.16 and 0.17?

**Comment by sunli829** [comment]
When I modify both of these, it will compile. 🙂

```toml
async-graphql = { git = "https://github.com/simplificAR/async-graphql", branch = "custom_handler", version ="1.17" }
```

```rust
rocket::ignite()
        .attach(GraphQL::fairing_with_playground(schema, "/", "/graphiql"))
```

**Comment by sunli829** [comment]
> Do you mean that this is a difference between 1.16 and 1.17?

It should be a problem caused by different dependencies. One is the git address and the other is the version number. Rustc will treat it as two different libraries. Maybe so. 😂

**Comment by Weasy666** [comment]
Great! Thanks for pointing that out 😀. Then i will add some docs to `async-graphql-rocket` and start a PR to merge the basic functionallity.

After that gets merged, i will start a PR to the examples repository, to add the starwars example. Should i also bump the version in the other examples?

After that, i would take a look at the streaming/upload stuff. Subscription with Websockets is currently not supported by Rocket, or the be more precise, it could be done with async Rocket, but there is currently no rocket nativ WS support.

Is this approach okay for you?

**Comment by sunli829** [comment]
Thank you for your contribution. It would be better if you could add the example of uploadfile, I have a rough look at your code, because I am not good at using the Rocket library, So I am not sure whether it can work properly.
Thank again!

**Comment by Weasy666** [comment]
You are welcome. Yeah, i obviously need to test it first 😅
Ok, then i will add uploadfile tomorrow and then do a PR.

**Comment by Weasy666** [comment]
Ok...i had to change my approach with the routes.

It is now compiling, see [here](https://github.com/simplificAR/async-graphql/tree/master/async-graphql-rocket), and the StarWars example is also working, see [here](https://github.com/simplificAR/examples/tree/master/rocket/starwars).

**Comment by sunli829** [comment]
Please change this path to `/`, consistent with other examples.

```rust
#[rocket::get("/")]
fn graphql_playground() -> content::Html<String> {
    content::Html(playground_source(GraphQLPlaygroundConfig::new("/")))
}
```

**Comment by Koxiaet** [comment]
This should probably be closed now, `async-graphql-rocket` isn't released but that's blocked on Rocket itself.
